File: lecture006/dutchFlag0123.cpp
//  use dutch flag to sort array with only 0,1,2,3

#include<iostream>
#include<vector>
using namespace std;

void input(vector <int> &arr) {
	for (int i = 0; i < arr.size(); i++) {
		cin>>arr[i];
	}
}

void display(vector <int> &arr) {
	for (int i : arr) {
		cout<<i<<" ";
	}
	cout<<endl;
}

int countThree(vector <int> &arr) {
	int count = 0;
	for (int i : arr) {
		if (i == 3) {
			count++;
		}
	}
	return count;
}


/*
	void swap(vector <int> arr, num1, num2) {
		int temp = arr[num1];
		arr[num1] = arr[num2];
		arr[num2] = temp;
	}
*/


void sortDutchFlag(vector <int> &arr) {
	int threes = countThree(arr);
	int ptr1 = 0, ptr2 = arr.size() - threes - 1, ptr3 = arr.size() - 1, itr = 0;
	while (itr < ptr2) {
		if (arr[itr] == 0) {
			swap(arr[itr], arr[ptr1]);
			ptr1++;
			itr++;
		}
		else if (arr[itr] == 2) {
			swap(arr[itr], arr[ptr2]);
			ptr2--;
		}
		else if (arr[itr] == 3) {
			swap(arr[itr], arr[ptr3]);
			ptr3--;
		}
		else {
			itr++;
		}
	}

}

int main(int args, char**argv) {
	cout<<"enter size: ";
	int size;
	cin>>size;
	cout<<"enter data in array: ";
	vector <int> arr(size);
	input(arr);
	// swap(arr[0],arr[arr.size() - 1]);
	display(arr);
	sortDutchFlag(arr);
	display(arr);
	return 0;
} 

File: lecture006/dutchFlag0123.java
// use dutch flag to sort array with only 0,1,2,3

import java.util.Scanner;

public class dutchFlag0123 {
	public static Scanner scn = new Scanner(System.in);
	
	public static void main(String[] args) {
		
		System.out.print("enter the size: ");
		int size = scn.nextInt();
		int[] arr = new int[size];
		input(arr);
		display(arr);
		sortDutchFlag(arr);
		display(arr);
	}

	public static void input(int[] arr) {
		for (int i = 0; i < arr.length; i++) {
			arr[i] = scn.nextInt();
		}
	}

	public static void display(int[] arr) {
		for (int i : arr) {
			System.out.print(i + " ");
		}
		System.out.println();
	}

	public static int countThree(int[] arr) {
		int count = 0;
		for (int i : arr) {
			if (i == 3) {
				count++;
			}
		}
		return count;
	}

	public static void swap(int[] arr, int num1, int num2) {
		int temp = arr[num1];
		arr[num1] = arr[num2];
		arr[num2] = temp;
	}

	public static void sortDutchFlag(int[] arr) {
		int threes = countThree(arr);
		int ptr1 = 0, ptr2 = arr.length - threes - 1 , ptr3 = arr.length - 1, itr = 0;
		while (itr <= ptr2) {
			if (arr[itr] == 0) {
				swap(arr, itr, ptr1);
				ptr1++;
				itr++;
			}
			else if (arr[itr] == 2) {
				swap(arr, itr, ptr2);
				ptr2--;
			}
			else if (arr[itr] == 3) {
				swap(arr, itr, ptr3);
				ptr3--;
			}
			else {
				itr++;
			}
		}
	}
}